File: swift5/Sources/swift5/generating/abstractFactory/MainAbstractFactory.swift

fileprivate enum CarType: String {
    case PassengerCar = "Passenger car"
    case SUV = "SUV"
}

fileprivate protocol Car {
    func getCarType() -> String;
}

fileprivate class PassengerCar : Car {
	public func getCarType() -> String {
		return CarType.PassengerCar.rawValue;
	}
}

fileprivate class SUV : Car {
	public func getCarType() -> String {
		return CarType.SUV.rawValue;
	}
}

fileprivate protocol AbstractFactory {
	func getTransmission(type: TransmissionType) -> Transmission?;
	func getCarType(value: CarType) -> Car?;
}

fileprivate class CarFactory : AbstractFactory {
    public func getCarType(value: CarType) -> Car? {
        switch value {
        case .PassengerCar:
			return PassengerCar();
		case .SUV:
			return SUV();
		default:
			return nil;
		} 
    }

    public func getTransmission(type: TransmissionType) -> Transmission? {
        return nil;
    }
}

fileprivate enum TransmissionType: String {
	case FiveGear = "5 gear"
	case SixGear = "6 gear"
	case SevenGear = "7 gear"
}

fileprivate protocol Transmission {
	func getType() -> String
}

class DefaultTransmission : Transmission {
	func getType() -> String {
		return TransmissionType.FiveGear.rawValue;
	}
}

class UpgratedTransmission : Transmission {
	func getType() -> String {
		return TransmissionType.SixGear.rawValue;
	}
}

class RaceTransmission : Transmission {
	func getType() -> String {
		return TransmissionType.SevenGear.rawValue;
	}
}

fileprivate class TransmissionFactory : AbstractFactory {
    public func getCarType(value: CarType) -> Car? {
        return nil
    }

    public func getTransmission(type: TransmissionType) -> Transmission? {
        switch type {
        case .SixGear:
			return UpgratedTransmission();
		case .SevenGear:
			return RaceTransmission();
		default:
			return DefaultTransmission();
		} 
    }
}

public class MainAbstractFactory {
	public static func main() {
		let carFactory: AbstractFactory = CarFactory();
		var car: Car? = carFactory.getCarType(value: CarType.PassengerCar);
		print(car!.getCarType());
		car = carFactory.getCarType(value: CarType.SUV);
		print(car!.getCarType());
	}
}

File: python/generating/abstractFactory/MainAbstractFactory.py

from enum import Enum
from abc import ABC, abstractmethod

class CarType(Enum):
    PassengerCar = "Passenger car"
    SUV = "SUV"

class TransmissionType(Enum):
    FiveGear = "5 gear"
    SixGear = "6 gear"
    SevenGear = "7 gear"

class Car(ABC):
    @abstractmethod
    def getCarType(self):
        pass

class PassengerCar(Car):
    def getCarType(self):
        return CarType.PassengerCar.value

class SUV(Car):
    def getCarType(self):
        return CarType.SUV.value

class Transmission(ABC):
    @abstractmethod
    def getType(self):
        pass

class DefaultTransmission(Transmission):
    def getType(self):
        return TransmissionType.FiveGear.value

class UpgratedTransmission(Transmission):
    def getType(self):
        return TransmissionType.SixGear.value

class RaceTransmission(Transmission):
    def getType(self):
        return TransmissionType.SevenGear.value

class AbstractFactory(ABC):
    @abstractmethod
    def getCarType(self, type: CarType):
        pass

    @abstractmethod
    def getTransmission(type):
        pass

class CarFactory(AbstractFactory):
    def getCarType(self, type):
        if (isinstance(type, CarType)):
            match type:
                case CarType.PassengerCar:
                    return PassengerCar()
                case CarType.SUV:
                    return SUV()
        else:
            return None

    def getTransmission(type):
        return None

class TransmissionFactory(AbstractFactory):
    def getCarType(self, type):
        return None

    def getTransmission(type):
        if (isinstance(type, TransmissionType)):
            match type:
                case TransmissionType.SixGear:
                    return UpgratedTransmission()
                case TransmissionType.SevenGear:
                    return RaceTransmission()
                case _:
                    return DefaultTransmission()
        else:
            return None

def main():
    abstractFactory = CarFactory()
    car = abstractFactory.getCarType(CarType.SUV)
    print(car.getCarType())

main()
